### ISSUE TITLE
Feature: Activity Error/Fault Handling

### ISSUE THREAD

**Issue by terry-delph** [elsa-workflows/elsa-core#200]
Hi @sfmskywalker,
 
Has any thought been put into handling errors/faults?

Could the **"ActivityBase"** class define a outcome of "Faulted" as standard, or change the below to always add OutcomeNames.Fault to the list.

```
public static ActivityDescriptor Describe(Type activityType)
        {
            var activityDefinitionAttribute = activityType.GetCustomAttribute<ActivityDefinitionAttribute>();
            var typeName = activityDefinitionAttribute?.Type ?? activityType.Name;
            
            var displayName =
                activityDefinitionAttribute?.DisplayName ??
                activityType.Name.Humanize(LetterCasing.Title);
            
            var description = activityDefinitionAttribute?.Description;
            var runtimeDescription = activityDefinitionAttribute?.RuntimeDescription;
            var category = activityDefinitionAttribute?.Category ?? "Miscellaneous";
            var icon = activityDefinitionAttribute?.Icon;
            var outcomes = activityDefinitionAttribute?.Outcomes ?? new[] { OutcomeNames.Done };

           //Insert OutcomeNames.Faulted into the list here

            var properties = DescribeProperties(activityType);

            return new ActivityDescriptor
            {
                Type = typeName.Pascalize(),
                DisplayName = displayName,
                Description = description,
                RuntimeDescription = runtimeDescription,
                Category = category,
                Icon = icon,
                Properties = properties.ToArray(),
                Outcomes = outcomes
            };
        }
```

If execution faulted, and the workflow has a connection for the "Faulted" outcome, follow its path to another activity and schedule it. If there is no connection, just fault the workflow as normal.

The next activity could just go back some steps and try the set of steps again (it could also do this a set number of times), or it could do a rollback to perform a clean up, then exit as faulted.

It could even point back to itself, to perform a retry with a TimedEvent activity.

Hopefully the above makes sense to you.

Thanks

**Comment by sfmskywalker** [comment]
Hi @terry-delph 

Many thanks for the above, that is quite good thinking. 

I am hesitant however to have **every** activity expose a **Faulted** outcome, because there are activities that are not expected to fault, like `WriteLine` and `SetVariable` (if they were to fault, that would be either a bug in the activity or some off-chance fluke where the host is crashing or something like that). I don't want to unnecessarily "clutter" the activity designer just because in theory there's a minute chance that any line of execution could fail.

I do realize that there are many activities that are expected to potentially fault, such as any network-related activities to name just one category.

Perhaps what we can do is this: an activity developer uses the `ActivityDefinitionAttribute` to specify possible outcomes. If the activity can fault, the developer should include `OutcomeNames.Faulted` as a possible outcome.

When the workflow invoker executes the activity, and the activity throws an exception, the workflow invoker checks if the activity exposes a "Faulted" outcome. If so, any connected activity to that outcome will be scheduled next, using he exception as an input for that activity.

Alternatively, or maybe additionally, an activity could implement a "HandleFault" method. The activity can return any ActivityExecutionResult it wants. Perhaps even a `RetryResult` to implement some kind of error policy.

Or we could leave error handling up to each activity itself: if an activity can fault, it could simply do a try/catch block themselves, and return the appropriate outcome (e.g. "Faulted").

Some more food for thought, not immediately relevant: In a future version, I am thinking about container activities, like ForEach, Sequence, While, etc. but also maybe a `Try` activity. Any activity faulting in that container would trigger the "Faulted" outcome of the "Try" activity.


As you can see, there are a few thoughts, but no concrete decisions yet :)
Any input on this is greatly appreciated.

**Comment by terry-delph** [comment]
> I am hesitant however to have every activity expose a Faulted outcome, because there are activities that are not expected to fault, like WriteLine and SetVariable (if they were to fault, that would be either a bug in the activity or some off-chance fluke where the host is crashing or something like that). I don't want to unnecessarily "clutter" the activity designer just because in theory there's a minute chance that any line of execution could fail.

That is well justified for specifically stating it can Fault rather than always.
Alternatively, the IActivity interface could have a boolean "CanFault" property that can be used for that purpose.
If true, automatically add the Faulted outcome.

> When the workflow invoker executes the activity, and the activity throws an exception, the workflow invoker checks if the activity exposes a "Faulted" outcome. If so, any connected activity to that outcome will be scheduled next, using he exception as an input for that activity.

Love the sound of that.

> Alternatively, or maybe additionally, an activity could implement a "HandleFault" method. The activity can return any ActivityExecutionResult it wants. Perhaps even a RetryResult to implement some kind of error policy.

Love this also, and makes the activity more configurable for different scenarios of transient errors.

> Some more food for thought, not immediately relevant: In a future version, I am thinking about container activities, like ForEach, Sequence, While, etc. but also maybe a Try activity. Any activity faulting in that container would trigger the "Faulted" outcome of the "Try" activity.

This would be great for grouping activities that need to be executed together after a failure and then retrying.
If will also be more visually pleasant and easier to understand if the designer houses them in a container too.

**Comment by Andale-online** [comment]
I'd like to see that system wide... With all of the features :-D
No rly:
- Exception handling is even one of my problems - since i use a lot of external code by reflection i do it in external code
- Would be great as a global logic for the "Fault Outcome" in the Activities Base Class
- An activity would NOT need to implement any exception handling - only override if needed (extended exceptions)
- Would upgrade all existing activities with the feature
- Overridden extended exception handling could even pass additional outcomes for seperate exception types
- Please: No Try Activity! Voting against that. You kept the framework straight and thats good. Exceptions are visible in the Dashboard. What happens as soon someone "Tr(y)ies" something: 90% of all exceptions are lost in /dev/null because of nobody logging :-( Make it global please
- Retry would be a great feature though - i tried to do that in logic, but i think the dashboard told me it would be cyclic referencing afair. This shouldnt be done in exception handling, sounds overkill to me, sorry. Maybe an activity "Retry" placed before the activity to retry with retry options and two outcomes (ToRetry and Fault) would be better. Whatta sentence, hope you get it. This would just hook INTO the exception handling (as e.g. an event with doRaiseException = false in event params or such).

Just my two bucks, dont shoot at me :-)

**Comment by craigfowler** [comment]
It seems that this is worth considering alongside #580 and vice-versa.  Even if they're not implemented at the same time.

**Comment by craigfowler** [comment]
I would actually advise _against_ a general "activity faulted" feature being a part of the Elsa.  The reason is that the semantics of this are very similar to `catch(Exception) { /* Pretend we can deal with every exception in the universe */ }`.

One of the great things about managed exceptions is that you may choose to catch only the exceptions which you expect and take remedial action specific to that exception.  Taking "Send HTTP request" as an example, it _could_ fail because any of the following reasons _or quite a few others_.  Of course, most developers only think about the first one or maybe the 2nd.

* The remote server returned a >= 400 status code
* The request timed out waiting for the remote server
* The remote server wasn't found on the network
* The app is running in an environment which doesn't have permission to make network requests

_Heck, even an activity which uses `Console.WriteLine` could fail if the console output has been redirected to a non-writable stream :fearful: .  Although anyone who actually does this deserves everything that's coming to them!_

In many ways I think the "right thing to do" genuinely is to have each activity implement its own exception-handling logic.  That can be done in a way that is appropriate to the ways in which that specific activity is expected to fail.  Yep, `try/catch` inside the activity class.  That includes catching _only the exceptions you expect and can deal with_, as appropriate.  The activity class knows exactly what code it's exercising and so it is the most suited class to know how to catch the right exceptions.

The way that an activity class "deals with" an exception _could legitimately be_ to return an appropriately-named result that indicates a specific well-known failure.  Back to the HTTP request example above, it could be written such that it could return outcomes such as `BadRequestFailure` or `ServerErrorFailure` or `TimeoutFailure` etc.

Subsequent activities in the workflow design _could_ then consume those specific outcomes using the standard mechanism for differing outcomes.  Back to the HTTP request example, I could totally foresee the use of an activity which implements _"Retry maximum X times with Y pause between retries"_ logic.  That could be used if the HTTP request responded with a timeout failure, but wouldn't be used if the HTTP request responded with a bad request failure (which isn't worth retrying).

As for unhandled exceptions which come out of activity methods, I honestly think that the best and only sane thing to do with these is:

* Log the exception in some manner
  * Do that using the global logging mechanism
  * _Perhaps_ also add the `.ToString()` exception to a property of the workflow itself
* Put the whole workflow instance into an error state and halt it indefinitely

In other words, treat unhandled exceptions as either coding errors or serious unforeseen problems.  If it's a serious unforeseen problem then we can't presume that we know how to deal with it.  In either case, the workflow may not continue.

**Comment by sfmskywalker** [comment]
I agree, let's not have a general Try/Catch activity to catch any and all exceptions, but instead let individual activities decide what to handle and how to expose outcomes (the HTTP Request activity is a great example).

**Comment by sfmskywalker** [comment]
We are tracking activity error outcomes here: #1115

**Comment by hakancunier** [comment]
May be late but i want you to focus on this use-case;
In onExecute event of a custom activity which blocks the workflow for user action performing, we call Suspend which is ASYNC. Everything in activity OnExecute works fine until here but somehow the suspend action fails how to catch this?

Also same case can be applied on OnResume event. We call Done method which is ALSO ASYNC and may throw an exception. I dont want my customers (and also not hoping) to implement a onerror activity planning, per their viewpoint the software must handle errors. So to achive this one must need to handle exceptions in an activity. And also for my design experience I would/wish implement an OnError event for both usability and readibility.

Can you re-think this behavior?

**Comment by sfmskywalker** [comment]
Hi @hakancunier ,

Thank you for taking the time to share your concern.

If I understand correctly, your main concern is: what should happen when a "system" call like "Done" or "Suspend" fails with an exception, is that correct?

If so, then I agree that it should be up to the software to handle these system-level errors, not the users designing the workflow.

**Comment by hakancunier** [comment]
Hi @sfmskywalker ,
Certainly yes. A "system" call like "Done" / "Suspend" must inform if it failed. At least Custom Activity or a different Activity BASE class should support this i think.

Thank you...

**Comment by hakancunier** [comment]
Will be an improvement on this issue?

**Comment by sfmskywalker** [comment]
Hi @hakancunier,

Thank you for clarifying your concerns further.

While I understand the importance of ensuring that system-level calls like "Done" and "Suspend" are robust, I must admit that, in practice, I haven't observed these calls failing often. Of course, as with any system component, there's always a non-zero probability of encountering issues.

I think that the onus shouldn't be on the user to handle these (as far as I understand) rare, systemic failures. Instead, Elsa should have mechanisms in place to manage them gracefully, such as implementing strategies to recover from uncommitted states.

To aid in our planning and ensure we're comprehensively addressing your concerns, could you provide one or two specific scenarios where you've observed or foresee "Done" or "Suspend" failing? Additionally, if you could shed light on the potential implications of such failures from your perspective, it would be immensely helpful.

Your insights will greatly assist us in formulating a strategy to improve and fortify these critical system calls.

**Comment by hakancunier** [comment]
Scenario is such as;

1. Create a custom activity extending the Elsa's Activity class. Do something committed to the database in OnResume
2. Dont define a workflowContextProviderRefresher in the workflow, or define but return null in loadasync.
3. Supply a context ID (ie:999999)
4. Create a WriteConsole activity as first step with text property saying Hello in Literal.
5. Create a CustomActivity as second step defined at step 1
6. Create WriteConsole activity as last step and 
6.1. Set Input as 'workflowContext.anyVarName.nonProp1' in JavaScript.
6.2. Set Text as JavaScript just: 'notFoundableVar'.

This will raise error on step 5 with only workflowexecution result, after calling the resume of workflow. The developer has no chance to handle it in the resume caller place since the workflow is not hardcoded. Also there were No compile error, no publishing error, that will be a runtime error only on the running instance. 

Finally, things committed in step 1 is not undoable.

**Comment by sfmskywalker** [comment]
Thank you @hakancunier for the detailed explanation, that is very helpful! We will give all this some more thought as we evolve Elsa 3.
